### ISSUE TITLE
volume services cause container creation to fail when running rootless containers

### ISSUE THREAD

**Issue by julian-hj** [cloudfoundry/garden-runc-release#133]
## Description
When we turn on rootless_containers using the experimental ops file in cf-deployment, volume mounts no longer work properly.

Container creation fails with an error like this:
```
{"timestamp":"2019-03-13T22:24:09.374325912Z","level":"error","source":"guardian","message":"guardian.create.containerizer-create.depot-create.generate-failed","data":{"error":"no such file or directory","handle":"deee3463-8ffe-409f-4d56-bcb4","path":"/var/vcap/data/garden/depot/deee3463-8ffe-409f-4d56-bcb4","session":"386.3.1"}}
```
The most obvious theory that would explain this behavior is that the "container root" user doesn't have access to the volume mount that diego places in `var/vcap/data/volumes/nfs/<volume-id>` but after some experimentation, that doesn't seem to be the case.

I manually mounted an NFS mount, and a mapfs mount on top of it, just as the nfs volume driver would do, and then created a "maximus" user and `su`'d to maximus.  As that user, I am able to interact with the share, write files, etc.

So...there's clearly something else going on...maybe some kind of containery namespacery thing that makes that path unavailable?  Or I suppose it could be complaining about some other path, although that seems a bit unlikely.

 
## Environment

* garden-runc-release version: the garden version is `4722aa257d136051bebe6df22f8f1cdce73fafb1`
* Enabled configuration: default cf-d + rootless_containers.yml
* IaaS: GCP
* Stemcell version: 250.9
* Kernel version: 4.15.0-45-generic

## Steps to reproduce
* from a cf-d deployment, add operations/enable-nfs-volume-service.yml, operations/test/enable-nfs-test-server.yml, and operations/experimental/rootless-containers.yml
* after deployment completes, `bosh run-errand nfsbrokerpush`
* log into cf and `cf enable-service-access nfs`
* `cf cs nfs Existing mynfs -c '{"share":"nfstestserver.service.cf.internal/export/user1000only", "version":"4", "uid":1000, "gid":1000}'`
* `cf push <some app> --no-start`
* `cf bind-service <some app> mynfs`
* `cf start <some app>`

Alternately, you might be able to reproduce this same issue in garden by just creating some kind of mount at `/var/vcap/data/volumes/nfs/79727b3d-7453-4c9f-97fd-5db014ef9fc8` and then add it to the list of bind mounts in the garden API to be mounted to `/var/vcap/data/79727b3d-7453-4c9f-97fd-5db014ef9fc8` in the container.  Since we get the same error with any kind of volume service (smb, nfs, mapfs) I am inclined to think that it has something to do with the layout of the rootfs, and isn't really about the mount at all.

## Logs

I dumped the dontpanic output into slack in this thread: https://cloudfoundry.slack.com/archives/C033RE5D6/p1552516723302700

## Cause

:man_shrugging: 

## Resolution

Yes, please!

**Comment by yulianedyalkova** [comment]
Hello @julian-hj!

Currently rootless garden is also running in a bpm container. In that container, we don't have access to the host filesystem, only to the directories provided as additional_volumes [here](https://github.com/cloudfoundry/garden-runc-release/blob/57bc50f5f88cf0d0f437d8c9a6ce88dbcf56e3ba/jobs/garden/templates/config/bpm.yml.erb#L11).

We do allow users to provide more additional volumes through  [this property](https://github.com/cloudfoundry/garden-runc-release/blob/57bc50f5f88cf0d0f437d8c9a6ce88dbcf56e3ba/jobs/garden/spec#L147-L149), so you should just add `/var/vcap/data/volumes/nfs` to the list when providing the nfs ops files.

**Comment by julian-hj** [comment]
@yulianedyalkova that makes sense, thanks.

We hit the same issue with Diego when the rep became a BPM job, and we fixed it there by generically adding /var/vcap/data/volumes to the list of additional volumes, but I guess we can fiddle with the garden properties in our ops files.

The one caveat though is that if Garden doesn't add the generic directory to its BPM profile, then this will definitely bite us on the backside if/when rootless becomes the default behavior and gets folded into PAS, since OpsMan doesn't use our ops files.  So, please either add the /var/vcap/data/volumes path to the set of additional directories, or give us the heads up when you are planning to run rootless.  Otherwise we are likely to have irate customers with broken volume services.

**Comment by julian-hj** [comment]
OK, after fiddling our ops files to add `/var/vcap/data/volumes/nfs` to `garden.additional_bpm_volumes` I am now seeing this error:

```
2019-04-01T23:22:54.27+0000 [API/0] OUT App instance exited with guid 9d21e5fc-1c64-426f-96b9-ec75725534a8 payload: {"instance"=>"a9f878d1-0e02-47d0-4b5f-a3a0", "index"=>0, "cell_id"=>"f373976b-368e-4680-b90b-48a99d6bc802", "reason"=>"CRASHED", "exit_description"=>"failed to create container: runc run: exit status 1: container_linux.go:345: starting container process caused \"process_linux.go:424: container init caused \\\"rootfs_linux.go:58: mounting \\\\\\\"/var/vcap/data/volumes/nfs/feefea09-4b7f-4644-905c-3036d37e2e4c-0108b40405e751f50c963477455d5a0d_a9f878d1-0e02-47d0-4b5f-a3a0\\\\\\\" to rootfs \\\\\\\"/var/vcap/data/grootfs/store/unprivileged/images/a9f878d1-0e02-47d0-4b5f-a3a0/rootfs\\\\\\\" at \\\\\\\"/var/vcap/data/grootfs/store/unprivileged/images/a9f878d1-0e02-47d0-4b5f-a3a0/rootfs/var/vcap/data/feefea09-4b7f-4644-905c-3036d37e2e4c\\\\\\\" caused \\\\\\\"operation not permitted\\\\\\\"\\\"\"\n", "crash_count"=>3, "crash_timestamp"=>1554160974226815464, "version"=>"0cf7c39d-a9a1-4fee-8fb5-08c5836abe1a"}
```
It should not be the case that this is happening because of posix permissions on the mount--that's more or less open to the world.

**Comment by danail-branekov** [comment]
Hey @julian-hj 
Finally garden managed to provide a fix, TL;DR in order applications to be able to see mounts within directories that are mounted in the applications containers those submounts must be shared so that they are propagated across all mount namespaces (see [man mount](http://man7.org/linux/man-pages/man8/mount.8.html)/shared subtrees for details). The fix would most probably be available with the next garden release. 
However, I would appreciate your preliminary feedback on the fix. In order to give it a try, you would need to deploy with 
* bpm-release v1.1.0 (currently not in latest cf-deployment)
* garden-runc-release, commit [ba9d17abf01165fbdcdcae460d17a427b3cfd2ca](https://github.com/cloudfoundry/garden-runc-release/commit/ba9d17abf01165fbdcdcae460d17a427b3cfd2ca) or later
* An ops file declaring `/var/vcap/data/volumes/nfs` as additional BPM volume:
```yml
---
- type: replace
  path: /instance_groups/name=diego-cell/jobs/name=garden/properties/garden/additional_bpm_volumes?/-
  value: "/var/vcap/data/volumes/nfs"
```

Thanks!

**Comment by georgethebeatle** [comment]
Hey @julian-hj 

The fix for the issue would be available with the upcoming garden-runc release (1.19.3). Required bpm version is already in cf-deployment. Closing the issue, feel free to reopen if the problem is not fixed on your side.

**Comment by julian-hj** [comment]
Hey, sorry for the slow feedback!

This is still not working.  We think that [this line](https://github.com/cloudfoundry/garden-runc-release/blob/v1.19.3/jobs/garden/templates/config/bpm.yml.erb#L24) in your BPM template is incorrect, because of [this logic in BPM](https://github.com/cloudfoundry/bpm-release/blob/49d8ede5fc240cc3f135071aaadf19081e038182/src/bpm/runc/adapter/adapter.go#L77-L79)

We fiddled with the rendered template to work around that, but ran into some other issues afterwards, so we aren't sure if perhaps there are more subtle problems.  It may make sense to try to  cross team pair to get this issue landed.

**Comment by danail-branekov** [comment]
@julian-hj Your explanation does make sense. Interesting, I do remember that with our fix the application was pushable and was able to start. Do you observe other symptoms we need to pay attention to?

**Comment by julian-hj** [comment]
@danail-branekov,

Sorry, it occurred to me last night that I hadn't really described the new symptoms.  Basically what's happening now is that the mountpoint is visible as a directory inside Garden's BPM container, but because the parent folder isn't set up as a shared mount, the new mount created by volume services doesn't propagate.

The end result of this is that when Garden sets up the bind mount of the shared volume, it just binds a directory inside /var/vcap/data/volumes/nfs.  If you `cf ssh` into your application container and run `mount` or `df` you'll see that the mounted volume (by default at /var/vcap/data/<some ugly guid>) is mounted to a local device, and not a mapfs fuse mount or an nfs mount.  You'll also see that it is empty, and doesn't have any of the contents of your remote share.

HTH,
-Julian

**Comment by danail-branekov** [comment]
Hey @julian-hj 
FYI, we opened bug https://github.com/cloudfoundry/bpm-release/issues/127 to BPM to support shared mount-only volumes

**Comment by danail-branekov** [comment]
Hi @julian-hj 

BPM closed cloudfoundry/bpm-release#127 and I gave the whole setup a try, I think we got it right this time :)
What I now see in the application container is:
```bash
vcap@c2ac1cb6-dcee-491d-7ee8-9a5f:~$ cat /proc/self/mountinfo | grep nfs
594 571 0:131 / /var/vcap/data/666628dc-bbbf-4e94-a059-221938b9f74c rw,nosuid,nodev,relatime master:6 - fuse.mapfs /var/vcap/data/volumes/nfs/666628dc-bbbf-4e94-a059-221938b9f74c-78b90853fc0745e8d48bbbdb75a9c616_c2ac1cb6-dcee-491d-7ee8-9a5f_mapfs rw,user_id=0,group_id=0,allow_other
```

I also created a file at the shared volume (`/var/vcap/data/666628dc-bbbf-4e94-a059-221938b9f74c/foobar`) and verified that the file is available on the `nfstestserver` VM:

```bash
nfstestserver/ef472ddc-9e2d-4bc3-ba1b-ceedfe8308a7:~# find / -name 'foobar'
/export/user1000only/foobar
/var/vcap/store/nfstestserver/export/user1000only/foobar
```

I am therefore closing this issue, please feel free to reopen if something is not right.

cc @xoebus

**Comment by xoebus** [comment]
Great, I'll release later today.
